cardinal/server/server_test.go

```diff
@@ -100,6 +101,98 @@ func makeTestTransactionHandler(t *testing.T, world *ecs.World, opts ...Option)
 	}
 }
 
+func setTestTimeout(t *testing.T, timeout time.Duration) {
+	if _, ok := t.Deadline(); ok {
+		// A deadline has already been set. Don't add an additional deadline.
+		return
+	}
+	success := make(chan bool)
+	t.Cleanup(func() {
+		success <- true
+	})
+	go func() {
+		select {
+		case <-success:
+			// test was successful. Do nothing
+		case <-time.After(timeout):
+			//assert.Check(t, false, "test timed out")
+			panic("test timed out")
+		}
+	}()
+}
+
+func TestShutDownViaMethod(t *testing.T) {
+	//setTestTimeout(t, 10*time.Second) // If this test is frozen then it failed to shut down, create failure with panic.
+	w := inmem.NewECSWorldForTest(t)
+	sendTx := ecs.NewTransactionType[SendEnergyTx, SendEnergyTxResult]("sendTx")
+	assert.NilError(t, w.RegisterTransactions(sendTx))
+	w.AddSystem(func(world *ecs.World, queue *transaction.TxQueue, _ *ecs.Logger) error {
+		return nil
+	})
+	assert.NilError(t, w.LoadGameState())
+	txh := makeTestTransactionHandler(t, w, DisableSignatureVerification())
+	resp, err := http.Get("http://localhost:4040/health")
+	assert.Equal(t, resp.StatusCode, 200)
+	ctx := context.Background()
+	w.StartGameLoop(ctx, 1*time.Second)
+	for !w.IsGameLoopRunning() {
+		//wait until game loop is running.
+		time.Sleep(1 * time.Millisecond)
+	}
+	gameObject := NewGameManager(w, txh.Handler)
+	err = gameObject.Shutdown() //Should block until loop is down.
+	assert.NilError(t, err)
+	assert.Assert(t, !w.IsGameLoopRunning())
+	for {
+		time.Sleep(500 * time.Millisecond)
+		_, err := http.Get("http://localhost:4040/health")
+		if err != nil {
+			break //server is supposed to be shutdown.
+		}
+	}
+
```

Comment:
can't you just do:

```go
	_, err = http.Get("http://localhost:4040/health")
	assert.Check(t, err != nil)
```

at this point?

Reply:
The for loop here will call the url multiple times. 

Sometimes there's no error. Sometimes there is an error. In both cases it is correct behavior. I am waiting for the server to shutdown so there is no incorrect behavior here. 

I am more using the timeout thing above to see if there's a problem.

Reply:
This is the thing that replaces `txh.IsServerRunning()` btw. Similar pattern.

Reply:
have you been able to generate the case where there isn't an error? i've ran it multiple times and always get an error - which makes sense to me, it seems to me this part of the code can't be reached unless the server has been shut down

Reply:
I think I see what you're getting at now. You're saying because we expect shutdown to block until everything stops running the server should always error. Yeah I see what you're saying.

Reply:
@technicallyty I changed lmk what you think.

Reply:
it should also work for the signal one too, as you spinlock on the atomic boolean:

```go
for w.IsGameLoopRunning() {
		time.Sleep(500 * time.Millisecond)
	}
```

after this, we should be able to assert that the handler is dead, as it is shut down before the game loop

Reply:
oh hm good point. If I checked it in a different order this wouldn't be the case. But yeah. ok I'll change it.

Reply:
changed @technicallyty